lib/vrl/stdlib/src/map_values.rs

```diff
@@ -0,0 +1,140 @@
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct MapValues;
+
+impl Function for MapValues {
+    fn identifier(&self) -> &'static str {
+        "map_values"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT | kind::ARRAY,
+                required: true,
+            },
+            Parameter {
+                keyword: "recursive",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "map object values",
+                source: r#"map_values({ "a": 1, "b": 2 }) -> |value| { value = value + 1 }"#,
+                result: Ok(r#"{ "a": 2, "b": 3 }"#),
+            },
+            Example {
+                title: "recursively map object values",
+                source: r#"map_values({ "a": 1, "b": [{ "c": 2 }, { "d": 3 }], "e": { "f": 4 } }, recursive: true) -> |value| { value = if is_integer(value) { int!(value) + 1 } else { value } }"#,
+                result: Ok(r#"{ "a": 2, "b": [{ "c": 3 }, { "d": 4 }], "e": { "f": 5 } }"#),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let recursive = arguments.optional("recursive");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(MapValuesFn {
+            value,
+            closure,
+            recursive,
+        }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()).or_array(Collection::any()),
+                variables: vec![Variable {
+                    kind: VariableKind::TargetInnerValue,
+                }],
+                output: Output::Kind(Kind::any()),
+                example: Example {
+                    title: "map object values",
+                    source: r#"map_values({ "one" : "one", "two": "two" }) -> |value| { upcase(value) }"#,
+                    result: Ok(r#"{ "one": "ONE", "two": "TWO" }"#),
+                },
```

Comment:
Do we benefit from having another example here? Could use `self.examples()[0]` otherwise